### PR TITLE
Adapt Swift type metadata lookup to τ_depth_index convention.

### PR DESCRIPTION
**Pull request by adrian-prantl**
This is the LLDB work to adapt the new compiler convention to use
depth and index instead of name to lookup type metadata artificial
variables.

<rdar://problem/38306926>

### CODE REVIEW

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift/pull/16937
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift/pull/16937
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift/pull/16937
@swift-ci test

File: source/Plugins/Language/Swift/SwiftLanguage.cpp [hunk 1]
```diff
@@ -872,7 +872,7 @@ SwiftLanguage::GetHardcodedSummaries() {
         return swift_metatype_summary_sp;
       }
       if (valobj.GetName().GetLength() > 12 &&
-          valobj.GetName().GetStringRef().startswith("$swift.type.") &&
+          valobj.GetName().GetStringRef().startswith(u8"$\u03C4_") &&
```
Comment (dcci) [inline source/Plugins/Language/Swift/SwiftLanguage.cpp]: Would be nice to store the unicode string to a variable, or add a comment for posterity

File: source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp [hunk 2]
```diff
@@ -119,8 +119,11 @@ lldb::TypeSP DWARFASTParserSwift::ParseTypeFromDWARF(const SymbolContext &sc,
   }
 
   if (!compiler_type && name) {
-    if (name.GetStringRef().startswith("$swift.") ||
-        name.GetStringRef().startswith(SwiftLanguageRuntime::GetCurrentMangledName("_TtBp").c_str())) {  // This is the RawPointerType, need to figure out its name from the AST.
+    if (name.GetStringRef().startswith("$") ||
+        name.GetStringRef().startswith(
+            SwiftLanguageRuntime::GetCurrentMangledName("_TtBp")
```
Comment (dcci) [inline source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp:124]: Should we switch to the new mangled name here? 
Also, isn't the second condition made redundant by the first? All the mangled names now start with `$S`

File: source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp [hunk 3]
```diff
@@ -320,10 +326,16 @@ void SwiftUserExpression::ScanContext(ExecutionContext &exe_ctx, Status &err) {
               log->Printf("    [SUE::SC] Argument name: %s",
                           template_arg_name.AsCString());
 
-            CompilerType concrete_type =
-                exe_ctx.GetProcessRef()
-                    .GetSwiftLanguageRuntime()
-                    ->GetConcreteType(frame, template_arg_name);
+            CompilerType concrete_type = template_arg_type;
```
Comment (dcci) [inline source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp]: This and the snippet above (line 295) are exactly the same. Can we factor the logic out in a common function?

File: source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp [hunk 4]
```diff
@@ -1262,6 +1277,9 @@ MaterializeVariable(SwiftASTManipulatorBase::VariableInfo &variable,
     VariableMetadataVariable *variable_metadata =
         static_cast<VariableMetadataVariable *>(variable.m_metadata.get());
 
+    // FIXME: It would be nice if we could do something like
```
Comment (dcci) [inline source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:1284]: Why we can't?

File: source/Target/SwiftLanguageRuntime.cpp [hunk 5]
```diff
@@ -2406,12 +2430,15 @@ Value::ValueType SwiftLanguageRuntime::GetValueType(
...
-        static_type_flags.AllClear(eTypeIsArchetype)) {
+        static_type_flags.AllClear(eTypeIsGenericTypeParam)) {
+      // FIXME: This branch is not covered by any testcases in the test suite.
       if (is_indirect_enum_case || static_type_flags.AllClear(eTypeIsBuiltIn))
         return Value::eValueTypeLoadAddress;
     }
   }
 
+  // Enabling this makes the inout_variables test hang.
+  //  return Value::eValueTypeScalar;
```
Comment (dcci) [inline source/Target/SwiftLanguageRuntime.cpp:2431]: Can you elaborate this one? Is this dead code? if so, can we remove it?

File: source/Target/SwiftLanguageRuntime.cpp [hunk 6]
```diff
@@ -2406,12 +2430,15 @@ Value::ValueType SwiftLanguageRuntime::GetValueType(
     }
 
     if (static_type_flags.AllSet(eTypeIsSwift | eTypeIsPointer) &&
-        static_type_flags.AllClear(eTypeIsArchetype)) {
+        static_type_flags.AllClear(eTypeIsGenericTypeParam)) {
+      // FIXME: This branch is not covered by any testcases in the test suite.
```
Comment (dcci) [inline source/Target/SwiftLanguageRuntime.cpp:2424]: Then should we remove it or add a test?

**Review by dcci — CHANGES_REQUESTED** [review]
Thanks for the great piece of work. I have some comments inline.

File: source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp (hunk 2 above)
Comment (adrian-prantl) [reply source/Plugins/SymbolFile/DWARF/DWARFASTParserSwift.cpp:124]: It's possible that this is just dead code.

AFAIK all remaining occurrences of _T types in LLDB are hard-coded typenames in the Objective-C Foundation framework that are reserved for Swifty interfaces vended by various of Foundation classes. In this particular case though, I am not so sure.

File: source/Target/SwiftLanguageRuntime.cpp (hunk 6 above)
Comment (adrian-prantl) [reply source/Target/SwiftLanguageRuntime.cpp:2424]: Yes. This is just something I noticed while working on this patch. I'll put an unreachable there and see if it sticks.

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift/pull/16937
@swift-ci test

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/swift/pull/16937
@swift-ci test

**Comment by adrian-prantl** [comment]
@dcci: Can you take another look?

**Comment by adrian-prantl** [comment]
@dcci: Can you take another look?

**Comment by dcci** [comment]
This looks good to me but we should really add a test for type reconstruction in the compiler.

**Comment by adrian-prantl** [comment]
I did a new type reconstruction test over in the last revision of apple/swift#16937.

**Pull request merged**